Added linalg.solve

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.solve`.

`linalg_solve_out` uses in-place operations on the provided result tensor.

I modified `apply_solve` to accept tensor of Int instead of std::vector, that way we can write a function similar to `linalg_solve_out` but removing the error checks and device memory synchronization.

In comparison to `torch.solve` this routine accepts 1-dimensional tensors and batches of 1-dim tensors for the right-hand-side term. `torch.solve` requires it to be at least 2-dimensional.

Ref. #42666

IvanYashchuk committed:
wip linalg.solve

IvanYashchuk committed:
linalg_solve_out improvements

IvanYashchuk committed:
Fix MAGMA error checks

IvanYashchuk committed:
Implemented linalg_solve_out_info

IvanYashchuk committed:
Add static to linalg_solve_out_info

IvanYashchuk committed:
Basic tests pass

IvanYashchuk committed:
Added test_solve_batched_non_contiguous test_solve_errors

IvanYashchuk committed:
dot -> matmul

IvanYashchuk committed:
More tests for out= variant

IvanYashchuk committed:
Always unsqueeze 1-dimensional rhs input

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Test autograd for linalg_solve

IvanYashchuk committed:
_linalg_solve_out_helper name must end with an underscore

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
Updated comment in linalg.h dot -> matmul

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Added zero batch dimensions test cases

IvanYashchuk committed:
Added OpInfo-based tests for linalg.solve

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -361,35 +371,33 @@ static void apply_solve(Tensor& b, Tensor& A, std::vector<int64_t>& infos) {
   auto self_working_copy = cloneBatchedColumnMajor(self);
   auto A_working_copy = cloneBatchedColumnMajor(A);
-  std::vector<int64_t> infos(batchCount(self), 0);
+  auto infos = at::empty({batchCount(self)}, self.options().dtype(kInt));
```
```suggestion
  auto infos = at::empty({std::max<int64_t>(1, batchCount(self))}, self.options().dtype(kInt));
```

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1129,18 +1123,32 @@ AT_ERROR("solve: MAGMA library not found in "
   auto self_working_copy = cloneBatchedColumnMajor(self);
   auto A_working_copy = cloneBatchedColumnMajor(A);
-  std::vector<int64_t> infos(batchCount(self), 0);
+  auto infos = at::empty({batchCount(self)}, self.options().dtype(kInt));
```
```suggestion
  auto infos = at::empty({std::max<int64_t>(1, batchCount(self))}, self.options().dtype(kInt));
```

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
Make sure that info has at least one integer for zero batch dimension case

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
No need to transform Tensor to std::vector there is an overload for that

IvanYashchuk committed:
No need to subclass OpInfo

IvanYashchuk committed:
Removed comment

IvanYashchuk commented:
@mruberry this PR is ready for review.

Similarly to https://github.com/pytorch/pytorch/pull/48261 Here I prepared the ground for future functions that don't do error checks to resolve issues like https://github.com/pytorch/pytorch/issues/31546
I changed `apply_solve` to work with Tensor of ints instead of `std::vector<int64_t>` for storing LAPACK/MAGMA error codes.

I fixed leading dimension parameters (`lda`) to LAPACK/MAGMA calls so that 0x0 matrices and 0xNxN batched matrices work correctly.

`_out` variant is implemented to reuse the provided memory in `result` and `linalg_solve` is implemented by calling to `linalg_solve_out`. The requirement for the `result`'s memory layout are strict though, it must be Fortran contiguous.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -413,6 +420,97 @@ std::tuple<Tensor&,Tensor&> solve_out(Tensor& solution, Tensor& lu, const Tensor
+  if (result.dim() > 2) {
+    batchCheckErrors(infos, "linalg_solve");
+  } else {
+    singleCheckErrors(infos.item().toInt(), "linalg_solve");
```
TODO: the failing test is likely to be caused by this line.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -349,8 +349,18 @@ template<> void lapackLuSolve<float>(char trans, int n, int nrhs, float *a, int
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ solve ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+/*
+Computes the solution to a system of linear equations
```
This is a great comment.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -361,35 +371,32 @@ static void apply_solve(Tensor& b, Tensor& A, std::vector<int64_t>& infos) {
   auto n = A.size(-2);
   auto nrhs = b.size(-1);
-  auto lda = std::max(int64_t{1}, n);
+  auto lda = std::max<int64_t>(1, n);
```
Nice cleanup.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1100,22 +1100,20 @@ AT_ERROR("solve: MAGMA library not found in "
   if (b.dim() == 2) {
     auto ipiv = at::empty({n}, at::kInt);
-    magma_int_t info = 0;
+    infos = infos.to(at::kCPU);  // magmaSolve requires infos tensor to live on CPU
```
cc @heitorschueroff for MAGMA cross-device tracking

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1164,18 +1158,32 @@ AT_ERROR("solve: MAGMA library not found in "
+    batchCheckErrors(infos_, "solve_cuda");
   } else {
-    singleCheckErrors(infos[0], "solve_cuda");
+    singleCheckErrors(infos_[0], "solve_cuda");
```
style is inconsistent here with other checks:

```
singleCheckErrors(infos.item().toInt(), "linalg_solve");
```

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1164,18 +1158,32 @@ AT_ERROR("solve: MAGMA library not found in "
   auto self_working_copy = cloneBatchedColumnMajor(self);
   auto A_working_copy = cloneBatchedColumnMajor(A);
-  std::vector<int64_t> infos(batchCount(self), 0);
+  auto infos = at::empty({std::max<int64_t>(1, batchCount(self))}, self.options().dtype(kInt));
```
So is this pretending that `apply_solve` can use `infos` on CUDA? 

What if we just create `infos` on the CPU to begin with?

IvanYashchuk replied:
For CUDA there are two different MAGMA calls `magmaSolve`([`magma_*gesv_gpu`](https://icl.cs.utk.edu/projectsfiles/magma/doxygen/group__magma__gesv.html)) and `magmaSolveBatched`([`magma_*gesv_batched`](https://icl.cs.utk.edu/projectsfiles/magma/doxygen/group__magma__gesv__batched.html)).
For single matrix solves MAGMA requires a pointer to an integer residing in CPU memory, while for batched input MAGMA requires a pointer to an array of integers residing in device memory. I consider the batched case to be more used and therefore `infos` should be created on the GPU by default.
So `apply_solve` does use `infos` directly on CUDA and without movement of data between devices (at least visible to us, there could be some sync inside the magma function) but only for the batched case.
I agree that we should probably create `infos` on the CPU for the single matrix solves now.
But in general for all other future linear algebra functions `infos` can and should be on GPU, but it depends on the particular calls to the underlying library.

mruberry replied:
Interesting. We'll have a chance to revisit this when we look at cross-device synchronization more closely. cc @heitorschueroff. This seems OK for now. 

Thanks for your detailed explanation, @IvanYashchuk.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -413,6 +420,97 @@ std::tuple<Tensor&,Tensor&> solve_out(Tensor& solution, Tensor& lu, const Tensor
+
+// Solves a system of linear equations matmul(input, x) = other in-place
+Tensor& linalg_solve_out(Tensor& result, const Tensor& input, const Tensor& other) {
+  auto infos = at::empty({0}, input.options().dtype(kInt));
```
Curious about handling of infos here after reviewing the CUDA section.

Should this keep infos on CPU through the entire process? Since infos isn't exposed to users  it seems like there's little risk of doing so. Looking forward to your thoughts on this, @IvanYashchuk.

mruberry commented on test/test_ops.py:
```diff
@@ -29,13 +29,14 @@ class TestOpInfo(TestCase):
-
-        # NOTE: only tests on first sample
-        sample = samples[0]
+        # sample_inputs can have a function for generating the input that doesn't work for specified dtype
```
We should probably fix the sample input generators so they don't throw runtime errors instead of catching their errors.

mruberry replied:
This is a really interesting point. The goal of this test is to verify that the function doesn't work on the given dtypes, but it makes sense that OpInfo implementers aren't designing sample input functions that support these dtypes!

This workaround seems OK for now, but please file an issue for this problem and add a comment citing the issue so we can revisit this later.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -400,6 +400,53 @@
+linalg.solve(input, other, *, out=None) -> Tensor
+
+Computes the solution ``x`` to the matrix equation ``matmul(input, x) = other``
+with a square matrix, or batches of such matrices, :attr:`input` and multiple right-hand side vectors :attr:`other`.
```
"and multiple right-hand side"...

Would it be more accurate to say "and one or more..."?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -400,6 +400,53 @@
+
+Supports input of ``float``, ``double``, ``cfloat`` and ``cdouble`` dtypes.
+
+.. note:: If :attr:`input` is a non-invertible matrix or non-square matrix then a RuntimeError will be thrown.
```
How does this work with batching? Is it if any matrix in the batch is singular that a runtime error will be thrown?

IvanYashchuk replied:
Yes, if at least one matrix is singular then a runtime error is thrown specifying the index of the first singular matrix.

IvanYashchuk replied:
https://github.com/pytorch/pytorch/pull/48456/files#diff-7e17421f32124016eb8de04dc2f445da5786a28355e1addc72b305466f590180R1864-R1870

mruberry replied:
Can we add that condition to this sentence like:

"If :attr:\'input\' is a non-square or non-invertible matrix, or a batch containing non-square matrices or one or more non-invertible matrices, then a RuntimeError will be thrown. "

IvanYashchuk replied:
Done in da42d0e.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -400,6 +400,53 @@
+
+.. note:: If :attr:`input` is a non-invertible matrix or non-square matrix then a RuntimeError will be thrown.
+.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+
```
The NumPy documentation has a callout to lstsq:

https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html?highlight=solve#numpy.linalg.solve

I guess we can't do this, however, because torch.lstsq != np.linalg.lstsq? When we add torch.linalg.lstsq we should remember to update this doc, however.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -400,6 +400,53 @@
+    >>> x = torch.linalg.solve(A, b)
+    >>> torch.allclose(A @ x, b)
+    True
+    >>> b = torch.rand(3)
```
Add a comment highlight the broadcasting here.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -262,6 +262,19 @@ def sample_inputs_addmm(self, device, dtype, requires_grad):
                                     low=None, high=None, 
                                     requires_grad=False))),) 
 
+def sample_inputs_linalg_solve(self, device, dtype, requires_grad=False):
```
Add a comment describing this function, in particular the type of input cases it's generating and why they're important. That is, why do these sample inputs "represent" or "cover" the range of interesting inputs ot the function?

mruberry reviewed (commented):
Hey @IvanYashchuk!

This is a great PR. I made a few pretty minor comments/suggestions, including one design question around the handling of the infos tensor. Looking forward to hearing your thoughts!

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Added description for sample_inputs_linalg_solve

IvanYashchuk committed:
Use batchCheckErrors(Tensor) overload instead of converting Tensor to std::vector

IvanYashchuk committed:
Fixed bug for batched input error checks

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
Added entry for solve to linalg.rst

IvanYashchuk committed:
Added link to the issue about test_unsupported_dtypes

IvanYashchuk committed:
Added aten_name to op test

IvanYashchuk committed:
Fix mypy error

Generated function signature was
def _linalg_solve_out_helper_(input: Tensor, input: Tensor, infos: Tensor) -> Tensor: ...
So 'self' from native_functions.yml declaration is transformed into 'input'.
The second argument is renamed to 'other'

IvanYashchuk commented:
@mruberry I updated this pull request.
OpInfo-based tests now use `sample_inputs_func=`.

About 'infos' Tensor handling: https://github.com/pytorch/pytorch/pull/48456#discussion_r535925679

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -483,6 +483,53 @@
+with a square matrix, or batches of such matrices, :attr:`input` and one or more right-hand side vectors :attr:`other`.
+If :attr:`input` is batched and :attr:`other` is not, then :attr:`other` is broadcasted
+to have the same batch dimensions as :attr:`input`.
+The resulting tensor has the same shape as broadcasted :attr:`other`.
```
"as broadcasted" -> "as the (possible broadcast) :attr:\`other\`"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -483,6 +483,53 @@
+
+Computes the solution ``x`` to the matrix equation ``matmul(input, x) = other``
+with a square matrix, or batches of such matrices, :attr:`input` and one or more right-hand side vectors :attr:`other`.
+If :attr:`input` is batched and :attr:`other` is not, then :attr:`other` is broadcasted
```
"is broadcasted" -> "is broadcast"

IvanYashchuk replied:
If the array is _broadcastable_, then it is _broadcast_ across the larger array. We do _broadcasting_ of the smaller array and the result is the _broadcasted_ array.

For A, b in Ax=b. If `A` is batched and `b` isn't -> `b` is _broadcasted_ to have the same batch dimensions `A`. `x` has the same shape as the (possibly _broadcasted_) `b`.

See https://numpy.org/doc/stable/reference/generated/numpy.broadcast.html#numpy.broadcast
> `index` current index in **broadcasted** result
> `shape` Shape of **broadcasted** result.
> `size` Total size of **broadcasted** result.

Maybe we shouldn't use this term here anyway? Since we don't really do the broadcasting but create internally full copies with the compatible shape. To the user, it looks like broadcasting though.

mruberry replied:
I appreciate your point and if you prefer "broadcasted" that's OK.

As best I can tell "broadcasted" seems to be a less common but acceptable past tense for the word "broadcast." Here's an article arguing for "broadcast" vs. "broadcasted":
https://www.merriam-webster.com/words-at-play/is-it-broadcast-or-broadcasted-forecast-or-forecasted-usage. I would prefer we follow the AP Stylebook's recommendation and avoid it in favor of the more common past tense for "broadcast."

IvanYashchuk replied:
Thank you for the link! I've changed it to "broadcast".

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -370,6 +371,38 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+        (1,) - same as () but explicit
+        (3,) - solve for 3 vectors.
+    Zeros in dimensions are edge cases in the implementation and important to test for in order to avoid unexpected crashes.
+    """
```
Add a "TODO" comment here to support b requiring grad, too; currently this is untested. This can be done in a follow-up PR.

IvanYashchuk replied:
Thanks for spotting that! Now `b` requires grad and tested.

mruberry commented:
> @mruberry I updated this pull request.
> OpInfo-based tests now use `sample_inputs_func=`.
> 
> About 'infos' Tensor handling: [#48456 (comment)](https://github.com/pytorch/pytorch/pull/48456#discussion_r535925679)

This is awesome, thanks @IvanYashchuk. There's some follow-up work but just a last few nits on this PR and then we're gtg. Would you rebase, make those small fixes, and then let's get this merged?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
Do gradcheck for 'b' input to linalg.solve as well

IvanYashchuk committed:
Changed docs: (possibly broadcasted) 'other'

IvanYashchuk commented:
@mruberry I resolved merge conflicts and updated slightly `sample_inputs_linalg_solve` to have `b` input tested with gradcheck as well.
I still in doubts about "broadcast" vs "broadcasted" change https://github.com/pytorch/pytorch/pull/48456#discussion_r543116666

IvanYashchuk committed:
Changed 'broadcasted' -> 'broadcast'

IvanYashchuk committed:
Updated the note on runtime error

IvanYashchuk commented:
@mruberry thank you for reviewing this PR! I think it's ready for merge.

mruberry reviewed (approved):
Nice work, @IvanYashchuk!

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -426,6 +433,99 @@ std::tuple<Tensor&,Tensor&> solve_out(Tensor& solution, Tensor& lu, const Tensor
+// Solves a system of linear equations matmul(input, x) = other in-place
+// LAPACK/MAGMA error codes are saved in 'infos' tensor, they are not checked here
+static Tensor& linalg_solve_out_info(Tensor& result, Tensor& infos, const Tensor& input, const Tensor& other) {
+  TORCH_CHECK(infos.scalar_type() == kInt,
```
Ack! My mistake, @IvanYashchuk. I forgot to ask about validating that the tensors are on the correct device. Is there a check that input and other are both on the same device?

IvanYashchuk replied:
It is checked by calling [`_linalg_broadcast_batch_dims`](https://github.com/pytorch/pytorch/pull/48456/files#diff-76df047bca5c2d7a540bdcc4023d68641c4f15638d83c5c874d833b03f9636b9R473) which includes a call to a function `linearSolveCheckInputs` that does this check https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebraUtils.h#L65-L68

mruberry replied:
Thanks for verifying.

mruberry commented:
This will hit a merge conflict:

```
AssertionError: linalg_solve.out takes out arguments and has to be written in the new style. Please add `use_c10_dispatcher: full` to your operator in native_functions.yaml and write the C++ implementation to take out arguments in the end.
```

We now have to write operators in a new way. @smessmer, is there a public link we can point people to to implement operators using the new style, or a suggestion for how they can continue to use the legacy style?

mruberry commented:
I think that annotating the native_functions.yaml entries this PR creates with 

```
use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
```

will address this merge conflict.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-solve

IvanYashchuk committed:
Use 'use_c10_dispatcher: hacky_wrapper_for_legacy_signatures'

IvanYashchuk commented:
I added `use_c10_dispatcher: hacky_wrapper_for_legacy_signatures`. It would be very helpful to have something to read on this topic, why it was introduced, and what's the "non-hacky" way to do things. I didn't find anything with a quick search.

mruberry commented:
Failures are unrelated. Thanks @IvanYashchuk! 

@smessmer should make a document public soon, but here's a snippet:

> If you're writing new operators, please use the new style. This means:
> Add a use_c10_dispatcher: full annotation to your operator in native_functions.yaml.
```
- func: myop.out(int a, Tensor(a!) out) -> Tensor(a!)
    use_c10_dispatcher: full
```
> and have the C++ implementation take out arguments in the end, i.e.
```
// "aten::myop.out(int a, Tensor(a!) out) -> Tensor(a!)"
// do this
Tensor& at::native::myop_out(int a, Tensor& out) {...}
 
// instead of this
Tensor& at::native::myop_out(Tensor& out, int a) {...}
```

As I understand it, this change is just aligning the order of arguments in native_functions.yaml with the C++ signatures.

mruberry commented:
Hey @IvanYashchuk, sorry but it looks like this hit a conflict with complex svd and pinverse. Would you just rebase this?

IvanYashchuk committed:
Merge branch 'master' into linalg-solve
